autopush_rs/src/util/ddb_helpers.rs

```diff
@@ -501,6 +566,241 @@ impl DynamoStorage {
+                    key: ddb_item! {
+                        uaid: s => uaid.clone()
+                    },
+                    ..Default::default()
+                })
+            },
+            |err: &DeleteItemError| {
+                matches!(err, &DeleteItemError::ProvisionedThroughputExceeded(_))
+            },
+        ).map_err(|_| "Error fetching user".into());
+        Box::new(response)
+    }
+
+    fn get_uaid(
+        ddb: Rc<Box<DynamoDb>>,
+        uaid: &Uuid,
+        router_table_name: &str,
+    ) -> MyFuture<GetItemOutput> {
+        let uaid = uaid.simple().to_string();
+        let table_name = router_table_name.to_string();
+        let response = retry_if(
+            move || {
+                ddb.get_item(&GetItemInput {
+                    table_name: table_name.clone(),
+                    consistent_read: Some(true),
+                    key: ddb_item! {
+                        uaid: s => uaid.clone()
+                    },
+                    ..Default::default()
+                })
+            },
+            |err: &GetItemError| matches!(err, &GetItemError::ProvisionedThroughputExceeded(_)),
+        ).map_err(|_| "Error fetching user".into());
+        Box::new(response)
+    }
+
+    fn register_user(
+        ddb: Rc<Box<DynamoDb>>,
+        user: &DynamoDbUser,
+        router_table: &str,
+    ) -> MyFuture<PutItemOutput> {
+        let item = user.to_attribs();
+        let router_table = router_table.to_string();
+        let user = user.clone();
+        let response: MyFuture<PutItemOutput> = {
+            let ddb_response = retry_if(
+                move || {
+                    let user = user.clone();
+                    let attr_values = hashmap!{
+                       ":router_type".to_string() => val!(S => user.router_type),
+                       ":connected_at".to_string() => val!(N => user.connected_at),
+                    };
+                    debug!("Registering user: {:?}", item);
+                    ddb.put_item(&PutItemInput {
+                        item: item.clone(),
+                        table_name: router_table.clone(),
+                        expression_attribute_values: Some(attr_values),
+                        condition_expression: Some(
+                            r#"(
+                            attribute_not_exists(router_type) or
+                            (router_type = :router_type)
+                        ) and (
+                            attribute_not_exists(node_id) or
+                            (connected_at < :connected_at)
+                        )"#.to_string(),
+                        ),
+                        return_values: Some("ALL_OLD".to_string()),
+                        ..Default::default()
+                    })
+                },
+                |err: &PutItemError| matches!(err, &PutItemError::ProvisionedThroughputExceeded(_)),
+            ).map_err(|e| format!("Error storing user record: {:?}", e).into());
```

Comment:
Similar to above using `chain_err` allows you to losslessly retain the original error context (this implementation may lose some info)